### ISSUE TITLE
Passive scalar issue with AMR

### ISSUE THREAD

**Issue by morganemacleod** [PrincetonUniversity/athena#365]
Hi Folks, 

I have an issue that I first noticed in my own problem and have since been able to reproduce (though to a somewhat less extreme extent) in the blast with hydro. 

It stems from using adaptive refinement with passive scalars.  I think that when meshblocks are de-refined, their boundaries seem to introduce incorrect values of the scalar concentration. 

To reproduce I modified the blast problem by introducing a single passive scalar. I commented the changes to blast.cpp (attached) with "MM": [blast_scalar.cpp.txt](https://github.com/PrincetonUniversity/athena-public-version/files/4181814/blast_scalar.cpp.txt) (rename -> .cpp) 

I configure with: 
<code>
python configure.py --prob blast_scalar --coord spherical_polar --flux hllc -mpi -hdf5 --hdf5_path /opt/local --cxx g++ --nscalars=1
</code>

And I run with the following input file, which increases the number of AMR levels and changes the threshold for refinement vs the default. 
[athinput.blast_sph.txt](https://github.com/PrincetonUniversity/athena-public-version/files/4181840/athinput.blast_sph.txt)
(again, remove the .txt)

I read into a python notebook as follows:
```
import matplotlib.pyplot as plt
import numpy as np
import athena_read as ar

%matplotlib inline

d = ar.athdf("/Users/morganmacleod/DATA/athenaruns/blast/scalar/Blast.out1.00001.athdf")

plt.figure(figsize=(8,8))
th_ind = int(d['rho'].shape[1]/2)
rr,pp = np.meshgrid(d['x1v'],d['x3v'])
plt.pcolormesh(rr*np.cos(pp),rr*np.sin(pp),d['r0'][:,th_ind,:])
plt.colorbar(label='r')
plt.xlabel('x')
plt.ylabel('y')
plt.axis('equal')
```

To generate the following:
![download-6](https://user-images.githubusercontent.com/22401914/74167662-914cf800-4bf6-11ea-95c1-4722b9d25964.png)
What I see here, is that in the interior of the "bubble" where the mesh has de-refined, along the boundaries there are some artifacts introduced. 

When I ran this with my own problem (which unfortunately requires other customizations of athena) I was encountering values of r-> 1e20 ! But I haven't been able to reproduce that extreme behavior in the blast problem -- I thought that this, based on the unmodified public repo, was probably a more useful case. 

With a fixed mesh, for example, we see no artifacts in the scalar concentration:
![download-7](https://user-images.githubusercontent.com/22401914/74168349-99596780-4bf7-11ea-968f-9dad25ce6dd4.png)

Thanks for the help & insight. 

Morgan

**Comment by atomsite** [comment]
Hi
I believe my issue (#390) has been merged with this one, I was asked to add my problem file and compile commands:

[athinput.wr140.txt](https://github.com/PrincetonUniversity/athena/files/7165872/athinput.wr140.txt)

The video in the issue page was from a previous version of my code, but current version is still exhbiting the same issue.
This is with mpi enabled, so configuration is:

`./configure.py --prob cwb --nscal 4 -mpi -hdf5`

For `GCC`. Run command is:

`mpirun bin/athena -i athinput.wr140.txt`

The issue is only for AMR, I have managed to get extremely good results with SMR, but as this work covers elliptic orbit systems, AMR is needed to keep the processing time down.

**Comment by atomsite** [comment]
Hi, sorry to be a bother but I was just wondering what the progress was on the fix?

**Comment by c-white** [comment]
This is not a solution, but as I've been deep-diving into passive scalars lately, I thought I'd mention that I see similar artifacts at SMR refinement boundaries (see below)

The only appear in the background atmosphere, where floors are activating all the time. I noticed that my problems diminish if I add the conserved passive scalar array as input to `ConservedToPrimitive`, where it is modified if and when variable inversion decides to modify the conserved density. You can imagine catastrophic failures where `u(IDN)` is reset to a floor for some reason, but `s` is left unchanged and large, so some future call to `PassiveScalarConservedToPrimitive` will set `r = s / u(IDN)` to a very large value.

This is a nontrivial change (there are many variable inversions throughout the code), and it doesn't completely fix things. It does make me wonder if there is an issue with hydro floors in prolongation buffers having unintended consequences for scalars.

![scalar](https://user-images.githubusercontent.com/2496675/148626509-1c86c670-0b92-4467-b563-cbe56271cebd.png)

**Comment by c-white** [comment]
Does #434 help this issue at all?

**Comment by atomsite** [comment]
Hi Chris
I have not checked, I've been extremely busy finishing my thesis
I managed to get it working using SMR, and gotten good results from it.
Though I would love to use AMR in the future so I will look at this more
closely after I have submitted my thesis.
Thank you very much!
Joseph

**Comment by felker** [comment]
I would love to personally put a monetary bounty on this issue, but I am not sure that there is a standard way to do this for non-security-related bugs?

And having left off debugging this in 2020 with https://github.com/PrincetonUniversity/athena/issues/365#issuecomment-729114935 I know how painful it might be and probably cannot offer enough of a reward to make it a decent hourly ROI 😄

**Comment by c-white** [comment]
Possibly related to #446.

**Comment by felker** [comment]
I just confirmed that @morganemacleod 's original reported problem is fixed as of e8f1ee5a5813b7edaf1ca42b69bf4108172d01f2 whereas afcd297a9b2b8e91ae06c0637911d49f8eba081d from two weeks ago still had the artifacts in the first plot. 

```
make clean; python configure.py --prob blast_scalar --flux hllc -hdf5 --cxx g++ --nscalars=1 --hdf5_path=/usr/local -mpi -h5double --coord spherical_polar; make -j8
cd bin
mpirun -np 4 ./athena -i athinput.blast_sph
```
(FYI at first I had forgotten to include `--coord spherical_polar`, which was necessary to reproduce the problem on the older versions of the code, of course).

**Comment by felker** [comment]
My 2D problem reproducer from https://github.com/PrincetonUniversity/athena/issues/365#issuecomment-729114935
still triggers some s>1.0 warnings, but only at one time step. And tons of r>1.0 warnings, but this check is in the variable inversion and could be in the prolongation buffers, etc. Here is how I compiled and ran:
```
make clean; python configure.py --prob blast_scalar --flux hllc -hdf5 --cxx g++ --nscalars=1 --hdf5_path=/Users/felker/hdf5-serial/ -h5double; make -j8
cd bin
./athena -i athinput.blast_sph_2d time/nlim=64 time/integrator=vl2
```


```
cycle=3 time=1.3094734504890235e-03 dt=4.0424159088331131e-04
Conserved scalar exceeds 1.0; s(0,0,3,5) = 1.0000000000000162e+00
Conserved scalar exceeds 1.0; s(0,0,4,5) = 1.0000000000000162e+00
Conserved scalar exceeds 1.0; s(0,0,3,2) = 1.0000000000000162e+00
Conserved scalar exceeds 1.0; s(0,0,4,2) = 1.0000000000000162e+00

Primitive scalar exceeds 1.0; r(0,0,4,6) = 1.0000000000000002e+00
Primitive scalar exceeds 1.0; r(0,0,0,6) = 1.0000000000000002e+00
Primitive scalar exceeds 1.0; r(0,0,4,2) = 1.0000000000000002e+00
...
cycle=51 time=2.2131674380044111e-02 dt=4.5981656791154666e-04
Primitive scalar exceeds 1.0; r(0,0,2,3) = 1.0530641412451951e+00
Primitive scalar exceeds 1.0; r(0,0,2,4) = 1.0031918086024150e+00
Primitive scalar exceeds 1.0; r(0,0,2,5) = 1.0000640586284255e+00
Primitive scalar exceeds 1.0; r(0,0,3,3) = 1.0486935542292231e+00
Primitive scalar exceeds 1.0; r(0,0,3,4) = 1.0028060516123298e+00
Primitive scalar exceeds 1.0; r(0,0,3,5) = 1.0000552799189093e+00
Primitive scalar exceeds 1.0; r(0,0,4,4) = 1.0022943683438719e+00
Primitive scalar exceeds 1.0; r(0,0,4,5) = 1.0000427307192501e+00
```

Just put some simple checks in the code, so perhaps these arent actually meaningful:
```
diff --git a/src/eos/eos_scalars.cpp b/src/eos/eos_scalars.cpp
index 806db06a..824c816d 100644
--- a/src/eos/eos_scalars.cpp
+++ b/src/eos/eos_scalars.cpp
@@ -53,6 +53,9 @@ void EquationOfState::PassiveScalarConservedToPrimitive(

           //r_n = (r_n < scalar_floor_) ? scalar_floor_ : r_n;
           //s_n = r_n * d;
+          if (r_n > 1.0) {
+            std::cout << "Primitive scalar exceeds 1.0; r(0," << k << "," << j << "," << i <<") = " << r_n << std::endl;
+          }
         }
       }
     }
diff --git a/src/scalars/add_scalar_flux_divergence.cpp b/src/scalars/add_scalar_flux_divergence.cpp
index 9bd04226..569d7718 100644
--- a/src/scalars/add_scalar_flux_divergence.cpp
+++ b/src/scalars/add_scalar_flux_divergence.cpp
@@ -91,6 +91,10 @@ void PassiveScalars::AddFluxDivergence(const Real wght, AthenaArray<Real> &s_out
 #pragma omp simd
         for (int i=is; i<=ie; ++i) {
           s_out(n,k,j,i) -= wght*dflx(n,i)/vol(i);
+          if (s_out(n,k,j,i) > 1.0) {
+            std::cout << "Conserved scalar exceeds 1.0; s(0," << k << "," << j << "," << i <<") = "
+                      << s_out(n,k,j,i) << std::endl;
+          }
         }
       }
     }
```
